### ISSUE TITLE
Warning Message Suppressed on Test Failure

### ISSUE THREAD

**Issue by ashishshuklabs** [nunit/nunit#1985]
I'm using Warn.If to validate a few values in a grid. The execution continues as expected when the condition fails, which is great. However, the warning message is suppressed if any other assert fails in the test later down the line. I can see the warning message and the test in Inconclusive state when there are no further failures in the test. This is exactly opposite of what is written up in the documentation i.e.  
"If the test subsequently fails, the warnings will be displayed along with the failure message or - in the case of Assert.Multiple - messages."

Here's what I see:
**On an Assertion Failure:** Only the message associated with the current assertion failure is reported, the test status is Failed and the Warning Message is suppressed. (First Picture)
**On Assertion Pass:** Warning Message is displayed and the Test Status becomes inconclusive.(Second Picture)

Please confirm if this is the expected behavior with Warnings? Attached screenshots for both the cases described above.
![on assertion failure](https://cloud.githubusercontent.com/assets/20329724/21997233/0fac230a-dc93-11e6-923c-43258eb42d50.PNG)
![on assertion pass](https://cloud.githubusercontent.com/assets/20329724/21997234/0fadca0c-dc93-11e6-9df1-f45ac1f69abe.PNG)

**Comment by CharliePoole** [comment]
I see you are using the VS adapter. There has been no release of the VS adapter that supports either warnings or multiple asserts, so unless you were building from source you couldn't see that information. I'll edit the documentation to make that clearer.

The latest release (3.6) of the framework supports both warnings and multiple assert failures. It provides information to the runner in the XML that it transmits. If you use an old runner, then that info won't be displayed. In the case of the console, we did a 3.6 release that supports both features. There is a 3.7 release of the adapter that we are working on, which will support it. (Note that the version numbers of the different separate software products are no longer being kept in sync and will continue to diverge.)

In the case of multiple assert failures, the framework creates a failure message for use by legacy runners, which shows all the relevant messages (but not stack traces) That's what you are seeing in your second example. In the case of warnings, there is really nothing that can be done for legacy runners, since the outcome of "Warning" is not known to older runners. To make matters worse, it's not known to Visual Studio either, so even with the 3.7 adapter, the output will not be perfect. However, at least you will see all the errors, failures and warnings listed!

**Comment by ashishshuklabs** [comment]
Thanks for clarifying. I'll try using console runner 3.6. BTW, Thanks for coming up with Multi Asserts and Warnings. It was badly needed. Great Work Charlie and Team!!!!!

**Comment by ashishshuklabs** [comment]
Charlie,
Based on your comments above, just need to clarify something. If I use console 3.6, the results should display the warning messages right? I just ran the same test (second example) and I still do not see the warning messages in the Result XML (attached). Am I missing something here?
Thanks,
Ashish


<?xml version="1.0" encoding="utf-8" standalone="no"?>

<test-results name="C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\bin\Expert.Billing.UIAutomation.dll" total="1" errors="0" failures="1" not-run="0" inconclusive="0" ignored="0" skipped="0" invalid="0" date="2017-01-16" time="21:49:52">
  <environment nunit-version="3.5.0.0" clr-version="4.0.30319.42000" os-version="Microsoft Windows NT 10.0.14393.0" platform="Win32NT" cwd="C:\Users\ashish.shukla\Downloads\NUnit.Console-3.6.0" machine-name="xxxxx" user="xxxxx" user-domain="xxxxxx" />
  <culture-info current-culture="en-NZ" current-uiculture="en-US" />
  <test-suite type="Assembly" name="C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\bin\Expert.Billing.UIAutomation.dll" executed="True" result="Failure" success="False" time="50.940" asserts="22">
    <properties>
      <property name="_PID" value="39712" />
      <property name="_APPDOMAIN" value="domain-" />
    </properties>
    <failure>
      <message><![CDATA[One or more child tests had errors]]></message>
      <stack-trace />
    </failure>
    <results>
      <test-suite type="TestSuite" name="Expert" executed="True" result="Failure" success="False" time="50.922" asserts="22">
        <failure>
          <message><![CDATA[One or more child tests had errors]]></message>
          <stack-trace />
        </failure>
        <results>
          <test-suite type="TestSuite" name="Billing" executed="True" result="Failure" success="False" time="50.918" asserts="22">
            <failure>
              <message><![CDATA[One or more child tests had errors]]></message>
              <stack-trace />
            </failure>
            <results>
              <test-suite type="TestSuite" name="UIAutomation" executed="True" result="Failure" success="False" time="50.915" asserts="22">
                <failure>
                  <message><![CDATA[One or more child tests had errors]]></message>
                  <stack-trace />
                </failure>
                <results>
                  <test-suite type="TestSuite" name="Tests" executed="True" result="Failure" success="False" time="50.911" asserts="22">
                    <failure>
                      <message><![CDATA[One or more child tests had errors]]></message>
                      <stack-trace />
                    </failure>
                    <results>
                      <test-suite type="TestFixture" name="BillingUnitTests" executed="True" result="Failure" success="False" time="50.905" asserts="22">
                        <failure>
                          <message><![CDATA[One or more child tests had errors]]></message>
                          <stack-trace />
                        </failure>
                        <results>
                          <test-case name="Expert.Billing.UIAutomation.Tests.BillingUnitTests.AlertsMultiValidationWithWarning" executed="True" result="Failure" success="False" time="50.893" asserts="22">
                            <categories>
                              <category name="Warning Test" />
                            </categories>
                            <failure>
                              <message><![CDATA[  No Rows found in [PrebillDataGrid]
  Expected: greater than 0
  But was:  0
]]></message>
                              <stack-trace><![CDATA[at Expert.Billing.UIAutomation.Grids.PrebillDataGrid..ctor() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Grids\PrebillDataGrid.cs:line 53
at Expert.Billing.UIAutomation.Tabs.PrebillList.get_PrebillListGrid() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Tabs\PrebillList.cs:line 61
at Expert.Billing.UIAutomation.Tests.BillingUnitTests.AlertsMultiValidationWithWarning() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Tests\BillingUnitTests.cs:line 575
]]></stack-trace>
                            </failure>
                          </test-case>
                        </results>
                      </test-suite>
                    </results>
                  </test-suite>
                </results>
              </test-suite>
            </results>
          </test-suite>
        </results>
      </test-suite>
    </results>
  </test-suite>
</test-results>

**Comment by ashishshuklabs** [comment]
<?xml version="1.0" encoding="utf-8" standalone="no"?>

<test-results name="C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\bin\Expert.Billing.UIAutomation.dll" total="1" errors="0" failures="1" not-run="0" inconclusive="0" ignored="0" skipped="0" invalid="0" date="2017-01-16" time="21:49:52">
  <environment nunit-version="3.5.0.0" clr-version="4.0.30319.42000" os-version="Microsoft Windows NT 10.0.14393.0" platform="Win32NT" cwd="C:\Users\ashish.shukla\Downloads\NUnit.Console-3.6.0" machine-name="xxxxx" user="xxxxx" user-domain="xxxxxx" />
  <culture-info current-culture="en-NZ" current-uiculture="en-US" />
  <test-suite type="Assembly" name="C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\bin\Expert.Billing.UIAutomation.dll" executed="True" result="Failure" success="False" time="50.940" asserts="22">
    <properties>
      <property name="_PID" value="39712" />
      <property name="_APPDOMAIN" value="domain-" />
    </properties>
    <failure>
      <message><![CDATA[One or more child tests had errors]]></message>
      <stack-trace />
    </failure>
    <results>
      <test-suite type="TestSuite" name="Expert" executed="True" result="Failure" success="False" time="50.922" asserts="22">
        <failure>
          <message><![CDATA[One or more child tests had errors]]></message>
          <stack-trace />
        </failure>
        <results>
          <test-suite type="TestSuite" name="Billing" executed="True" result="Failure" success="False" time="50.918" asserts="22">
            <failure>
              <message><![CDATA[One or more child tests had errors]]></message>
              <stack-trace />
            </failure>
            <results>
              <test-suite type="TestSuite" name="UIAutomation" executed="True" result="Failure" success="False" time="50.915" asserts="22">
                <failure>
                  <message><![CDATA[One or more child tests had errors]]></message>
                  <stack-trace />
                </failure>
                <results>
                  <test-suite type="TestSuite" name="Tests" executed="True" result="Failure" success="False" time="50.911" asserts="22">
                    <failure>
                      <message><![CDATA[One or more child tests had errors]]></message>
                      <stack-trace />
                    </failure>
                    <results>
                      <test-suite type="TestFixture" name="BillingUnitTests" executed="True" result="Failure" success="False" time="50.905" asserts="22">
                        <failure>
                          <message><![CDATA[One or more child tests had errors]]></message>
                          <stack-trace />
                        </failure>
                        <results>
                          <test-case name="Expert.Billing.UIAutomation.Tests.BillingUnitTests.AlertsMultiValidationWithWarning" executed="True" result="Failure" success="False" time="50.893" asserts="22">
                            <categories>
                              <category name="Warning Test" />
                            </categories>
                            <failure>
                              <message><![CDATA[  No Rows found in [PrebillDataGrid]
  Expected: greater than 0
  But was:  0
]]></message>
                              <stack-trace><![CDATA[at Expert.Billing.UIAutomation.Grids.PrebillDataGrid..ctor() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Grids\PrebillDataGrid.cs:line 53
at Expert.Billing.UIAutomation.Tabs.PrebillList.get_PrebillListGrid() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Tabs\PrebillList.cs:line 61
at Expert.Billing.UIAutomation.Tests.BillingUnitTests.AlertsMultiValidationWithWarning() in C:\TFS\ExpertSuite\Dev\TestAutomation\Modules\Tests.TestStack\Expert.UIAutomation.Billing\src\Tests\BillingUnitTests.cs:line 575
]]></stack-trace>
                            </failure>
                          </test-case>
                        </results>
                      </test-suite>
                    </results>
                  </test-suite>
                </results>
              </test-suite>
            </results>
          </test-suite>
        </results>
      </test-suite>
    </results>
  </test-suite>
</test-results>
[NUnitTestResult.txt](https://github.com/nunit/nunit/files/709481/NUnitTestResult.txt)

**Comment by ashishshuklabs** [comment]
[NUnitTestResult.txt](https://github.com/nunit/nunit/files/709487/NUnitTestResult.txt)
![part1](https://cloud.githubusercontent.com/assets/20329724/22000767/f068bf36-dca5-11e6-8613-2e07e01fe0ff.PNG)
![part2](https://cloud.githubusercontent.com/assets/20329724/22000768/f069b2ec-dca5-11e6-84ae-64c460d97377.PNG)

**Comment by CharliePoole** [comment]
Your XML result `environment` element indicates you are using NUnit 3.5.

**Comment by CharliePoole** [comment]
Make sure you are referencing the 3.6 framework and running with the 3.6 console, which was just released.

**Comment by ashishshuklabs** [comment]
Hi Charlie,

Thanks for your valuable input.I just figured out that the warning messages
were getting suppressed as the test failed on Assert.That statement that
followed Warn.If in a previous step. Based on what I see, Warnings are
properly reported in case you use Assert.Multiple rather than just a plain
Assert.That. I just wrapped this assert within Assert.Multiple and there we
go, I see all the messages along with the failed Assertion. Based on what
I've seen so far, Warnings will be correctly reported when you're using
Assert.Multiple and might be a little flaky with plain asserts. Is this the
right conclusion?

Thanks,
Ashish
